gui/test/app/StanSampler/useStanSampler.test.ts

```diff
@@ -0,0 +1,250 @@
+// @vitest-environment jsdom
+
+import { expect, test, describe, vi, afterEach, onTestFinished } from "vitest";
+import "@vitest/web-worker";
+import { renderHook, waitFor, act } from "@testing-library/react";
+import useStanSampler, {
+  useSamplerOutput,
+  useSamplerProgress,
+  useSamplerStatus,
+} from "../../../src/app/StanSampler/useStanSampler";
+import { defaultSamplingOpts } from "../../../src/app/StanSampler/StanSampler";
+
+import fakeURL from "./empty.js?url";
+import erroringURL from "./fail.js?url";
+const erroringSamplingOpts = {
+  ...defaultSamplingOpts,
+  num_chains: 999,
+};
+
+const mockedStdout = vi
+  .spyOn(console, "log")
+  .mockImplementation(() => undefined);
+const mockedStderr = vi
+  .spyOn(console, "error")
+  .mockImplementation(() => undefined);
+
+vi.mock("tinystan", async (importOriginal) => {
+  const mockedLoad = await import("./MockStanModel");
+  const mod = await importOriginal<typeof import("tinystan")>();
+  mod.default.load = mockedLoad.default;
+  return mod;
+});
+
+afterEach(() => {
+  vi.clearAllMocks();
+});
+
+const loadedSampler = async () => {
+  const ret = renderHook(() => useStanSampler(fakeURL));
+  const status = renderHook(() => useSamplerStatus(ret.result.current.sampler));
+  await waitFor(() => {
+    expect(status.result.current.status).toBe("loaded");
+  });
+
+  onTestFinished(() => {
+    expect(ret.result.current.sampler?.status).toEqual(
+      status.result.current.status,
+    );
+
+    expect(mockedStdout).not.toHaveBeenCalledWith("terminating model worker");
+    ret.unmount();
+    expect(mockedStdout).toHaveBeenCalledWith("terminating model worker");
+  });
+
+  return [ret, status] as const;
+};
+
+describe("useStanSampler", () => {
+  test("empty URL should return undefined", () => {
+    const { result } = renderHook(() => useStanSampler(undefined));
+
+    expect(result.current.sampler).toBeUndefined();
+  });
+
+  test("other URLs are nonempty", async () => {
+    const { result } = renderHook(() => useStanSampler(fakeURL));
+
+    expect(result.current.sampler).toBeDefined();
+  });
+
+  describe("useSamplerStatus", () => {
+    test("loading changes status", async () => {
+      const { result, rerender } = renderHook(
+        (props: { url: string } | undefined) => useStanSampler(props?.url),
+        { initialProps: undefined },
+      );
+
+      const { result: statusResult, rerender: rerenderStatus } = renderHook(
+        (sampler) => useSamplerStatus(sampler),
+        { initialProps: result.current.sampler },
+      );
```

Comment:
Seems like you're going to be doing this a lot; consider breaking out into a function? Or at least giving an alias to the props type?